Games: Fix launching disk images and zip files

garbear opened the pull request:
## Description
This PR fixes two errors when launching games. Closes #15841.

### First fix

The first fix is for disk images. In `system/playercorefactory.xml` is the line:

```xml
<rule name="dvdimage" dvdimage="true" player="VideoPlayer" />
```

Setting `dvdimage="true"` calls into FileItem [here](https://github.com/xbmc/xbmc/blob/f4d6b4022/xbmc/FileItem.cpp#L1087) to see if the extension is a disk image: 

```c++
bool CFileItem::IsDiscImage() const
{
  return URIUtils::HasExtension(m_strPath, ".img|.iso|.nrg|.udf");
}
```

We want to avoid using VideoPlayer for fileitems with a game info tag, so I've modified the xml line to be:

```xml
<rule name="dvdimage" dvdimage="true" game="false" player="VideoPlayer" />
```

This skips the `IsDiscImage()` comparison for FileItems with a game info tag.

### Second fix

The second fix is for zip and 7z games not launching. They aren't matched to a player and default to VideoPlayer. When launched, VideoPlayer shows a buffering window then immediately exits to the previous window.

I fixed this by giving ListItems a game info tag in MyGames. Now, when no rules match, RetroPlayer is used to open the file.

## Motivation and Context

See #15841

## How Has This Been Tested?

Test builds: https://github.com/garbear/xbmc/releases/tag/retroplayer-18.5-20200127

### Launching a .zip file

Before:

```
DEBUG: CPlayerCoreFactory::GetPlayers(/Users/xxx/Library/Application Support/Kodi/userdata/addon_data/plugin.program.iagl/temp_iagl/88games.7z)
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: system rules
DEBUG: CPlayerSelectionRule::GetPlayers: matches rule: system rules
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: mms/udp
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: lastfm/shout
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: rtmp
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: rtsp
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: streams
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: dvd
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: sdp/asf
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: nsv
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: radio
DEBUG: CPlayerCoreFactory::GetPlayers: matched 0 rules with players
DEBUG: CPlayerCoreFactory::GetPlayers: adding videodefaultplayer (VideoPlayer)
DEBUG: CPlayerCoreFactory::GetPlayers: for video=1, audio=0
DEBUG: CPlayerCoreFactory::GetPlayers: for video=1, audio=1
DEBUG: CPlayerCoreFactory::GetPlayers: added 1 players
```

After:

```
Debug Print: CPlayerCoreFactory::GetPlayers(C:\Users\xxx\Desktop\temp_iagl\88games.zip)
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: system rules
Debug Print: CPlayerSelectionRule::GetPlayers: matches rule: system rules
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: mms/udp
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: lastfm/shout
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: rtmp
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: rtsp
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: streams
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: dvd
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: dvdimage
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: sdp/asf
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: nsv
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: radio
Debug Print: CPlayerCoreFactory::GetPlayers: matched 0 rules with players
Debug Print: CPlayerCoreFactory::GetPlayers: adding retroplayer
Debug Print: CPlayerCoreFactory::GetPlayers: added 1 players
...
Debug Print: Select game client dialog: Found 9 candidates
Debug Print: Adding game.libretro.cap32 as a candidate
Debug Print: Adding game.libretro.fbalpha2012 as a candidate
Debug Print: Adding game.libretro.mame2000 as a candidate
Debug Print: Adding game.libretro.mame2010 as a candidate
Debug Print: Adding game.libretro.hatari as a candidate
Debug Print: Adding game.libretro.uae as a candidate
Debug Print: Adding game.libretro.vice as a candidate
Debug Print: Adding game.libretro.reicast as a candidate
Debug Print: Adding game.libretro.yabause as a candidate
Debug Print: Select game client dialog: Found 1 installable clients
Debug Print: Adding game.libretro.fbalpha as an installable client
Debug Print: Select game client dialog: Loading savestate metadata
Debug Print: Loading savestate from C:\Users\xxx\Desktop\temp_iagl\88games.sav
Debug Print: Failed to open savestate file C:\Users\xxx\Desktop\temp_iagl\88games.sa
```

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

garbear committed:
Fix error playing disk image games

Game items ending in disk image extensions are currently played using
VideoPlayer instead of RetroPlayer. This results in a brief buffering
dialog, and then the player exits to the previous window.

garbear committed:
Fix launching zip and 7z files from MyGames

da-anda commented:
while I don't mind having an additional game condition possibility here, I think the logic behind `dvdimage=true` needs to be adjusted instead. It clearly says **DVD**image and not **DISC**image and should certainly return `false` for anything but DVDs (which also excludes blu-rays IMO). So the additional game condition should not be needed here at all.

Rechi commented on xbmc/games/windows/GUIWindowGames.cpp:
```diff
@@ -265,6 +265,14 @@ bool CGUIWindowGames::GetDirectory(const std::string &strDirectory, CFileItemLis
     items.SetContent(content);
 
+  // Ensure files have a game info tag so that they're recognized as games
+  for (auto it = items.begin(); it != items.end(); ++it)
```
use a range based for loop

garbear replied:
`items` is a `CFileItemList`, I tried to figure out how to range-based loop over it but couldn't. Do you know anywhere this is done?

Rechi replied:
I don't know if it is already done anywhere else, but if you look at
https://github.com/xbmc/xbmc/blob/f4d6b4022fc3122471f0d39aeb58abafc9412a10/xbmc/FileItem.h#L761-L766
and
https://github.com/xbmc/xbmc/blob/f4d6b4022fc3122471f0d39aeb58abafc9412a10/xbmc/FileItem.h#L598
you'll come to the following result.
```cpp
  for (const CFileItemPtr& item : items)
  {
    if (!item->m_bIsFolder)
      item->GetGameInfoTag();
  }

```

garbear replied:
updated

garbear commented:
> while I don't mind having an additional game condition possibility here, I think the logic behind dvdimage=true needs to be adjusted instead. It clearly says DVDimage and not DISCimage and should certainly return false for anything but DVDs (which also excludes blu-rays IMO). So the additional game condition should not be needed here at all.

Hard to change the logic without breaking anything. What if we change the name from `dvdimage` to `diskimage`?

jjd-uk commented:
I think there's a good chance it was called DVDimage simply because the player was called DVDPlayer up until fernet changed the name.

da-anda commented:
renaming it to `diskimage` would also be fine (while keeping the old name for backwards compatibility for at least 1 release). Would at least be way less confusing.

garbear committed:
Rename "dvdimage" selection tag to "discimage"

The prior behavior was kept as a fallback. If the fallback is used, a
deprecation warning is logged.

garbear commented:
@da-anda I renamed to `discimage` and added a deprecation warning

lrusak commented:
@garbear you have a build failure :)

```
08:16:39 /home/jenkins/workspace/LINUX-64/xbmc/games/windows/GUIWindowGames.cpp: In member function ‘virtual bool KODI::GAME::CGUIWindowGames::GetDirectory(const string&, CFileItemList&)’:
08:16:39 /home/jenkins/workspace/LINUX-64/xbmc/games/windows/GUIWindowGames.cpp:269:29: error: binding reference of type ‘CFileItemPtr& {aka std::shared_ptr<CFileItem>&}’ to ‘const std::shared_ptr<CFileItem>’ discards qualifiers
08:16:39    for (CFileItemPtr& item : items)
08:16:39                              ^~~~~
```

garbear commented:
@Rechi I modified the comment:

```
// Ensure a game info tag is created so that files are recognized as games
```

I need points to mutable CFileItems. It looks like the iters you linked are all const. Should I add non-const iterators, or is there a way to iterate over mutable items?

Rechi commented:
There seems to be something wrong with the (c)begin and (c)end functions of the CFileItemList class. The following should be the fix.
```diff
--- a/xbmc/FileItem.h
+++ b/xbmc/FileItem.h
@@ -760,8 +760,8 @@
 
-  VECFILEITEMS::const_iterator begin() { return m_items.cbegin(); }
-  VECFILEITEMS::const_iterator end() { return m_items.cend(); }
+  VECFILEITEMS::iterator begin() { return m_items.begin(); }
+  VECFILEITEMS::iterator end() { return m_items.end(); }
   VECFILEITEMS::const_iterator begin() const { return m_items.begin(); }
   VECFILEITEMS::const_iterator end() const { return m_items.end(); }
-  VECFILEITEMS::const_iterator cbegin() const { return m_items.begin(); }
-  VECFILEITEMS::const_iterator cend() const { return m_items.end(); }
+  VECFILEITEMS::const_iterator cbegin() const { return m_items.cbegin(); }
+  VECFILEITEMS::const_iterator cend() const { return m_items.cend(); }
 private:
```

Your actual issue is that you use `for (CFileItemPtr& item : items)` instead of `for (const CFileItemPtr& item : items)`, like in my code suggestion.
`CFileItemPtr` == `std::shared_ptr<CFileItem>`, so it's `for (const std::shared_ptr<CFileItem>& item : items)`, which means that only the shared pointer, but not the `CFileItem` itself, is const and the correct (non const) variant of `GetGameInfoTag` gets called.

garbear commented:
That looked a little strange to me, I'll include the fix in this PR. And you make a good point, I'll add const

garbear committed:
FileItem: Fix range iterator constness

garbear commented:
Updated

garbear commented:
jenkins build this please
